File: c++/c16_1_topological_sorting.cpp
#include <bits/stdc++.h>
#include <gtest/gtest.h>
using namespace std;
using ll = long long;
#include "c16_1_topological_sorting.h"

namespace {
	void test(const vector<vector<ll>>& adj) {
		ll n = ssize(adj);
		auto ret = topological(adj);
		vector<ll> pos(n);
		for (auto [i, a] : ret | views::enumerate)
			pos[a] = i;
		for (ll a = 1; a < n; ++a) {
			for (ll b : adj[a]) {
				EXPECT_LT(pos[a], pos[b]);
			}
		}
		ranges::sort(ret);
		EXPECT_EQ(ret, views::iota(1ll, n) | ranges::to<vector>());
	}
}

vector<ll> topological(const vector<vector<ll>>& adj) {
	ll n = ssize(adj);
	vector<ll> visited(n), ret;
	auto dfs = [&](this auto self, ll s) {
		if (visited[s])
			return visited[s] != 1;
		visited[s] = 1;
		if (!ranges::all_of(adj[s], self))
			return false;
		ret.push_back(s);
		visited[s] = 2;
		return true;
		};
	for (ll x = 1; x < n; ++x) {
		if (!visited[x] && !dfs(x)) {
			return {};
		}
	}
	ranges::reverse(ret);
	return ret;
}

TEST(C161TopologicalSorting, topological) {
	test({ {} });
	test({ {},{2},{3},{6},{1,5},{2,3},{} });
	EXPECT_EQ(topological({ {},{2},{3},{5,6},{1,5},{2},{} }), vector<ll>{});
}

File: python/c16_1_topological_sorting.py
import unittest
import sys

sys.setrecursionlimit(max(sys.getrecursionlimit(),1<<20))

def topological(adj):
    n=len(adj)
    visited=[0]*n
    ret=[]
    def dfs(s):
        if visited[s]:
            return visited[s]!=1
        visited[s]=1
        if not all(map(dfs,adj[s])):
            return False
        ret.append(s)
        visited[s]=2
        return True
    for x in range(1,n):
        if not visited[x] and not dfs(x):
            return None
    ret.reverse()
    return ret

class Test(unittest.TestCase):
    def assertTopological(self,adj):
        with self.subTest(adj=adj):
            n=len(adj)
            ret=topological(adj)
            self.assertCountEqual(ret, range(1,n))
            pos={v:i for i,v in enumerate(ret)}
            for a in range(1,n):
                for b in adj[a]:
                    self.assertLess(pos[a],pos[b])
    def test(self):
        self.assertTopological([[]])
        self.assertTopological([[],[2],[3],[6],[1,5],[2,3],[]])
        self.assertIsNone(topological([[],[2],[3],[5,6],[1,5],[2],[]]))

if __name__=='__main__':
    unittest.main()
